core/transaction-pool/graph/src/pool.rs

```diff
@@ -286,13 +244,204 @@ impl<B: ChainApi> Pool<B> {
+		}
+	}
+
+	/// Remove from the pool.
+	pub fn remove_invalid(&self, hashes: &[ExHash<B>]) -> Vec<TransactionFor<B>> {
+		self.data.remove_invalid(hashes)
+	}
+
+	/// Get an iterator for ready transactions ordered by priority
+	pub fn ready(&self) -> impl Iterator<Item=TransactionFor<B>> {
+		self.data.0.pool.read().ready()
+	}
+
+	/// Returns pool status.
+	pub fn status(&self) -> base::Status {
+		self.data.0.pool.read().status()
+	}
+
+	/// Returns transaction hash
+	pub fn hash_of(&self, xt: &ExtrinsicFor<B>) -> ExHash<B> {
+		self.data.0.api.hash_and_length(xt).0
+	}
+}
+
+impl<B: ChainApi> Clone for SharedPoolData<B> {
+	fn clone(&self) -> Self {
+		Self(self.0.clone())
+	}
+}
+
+impl<B: ChainApi> SharedPoolData<B> {
+	/// Imports a bunch of unverified extrinsics to the pool
+	pub fn submit_at<T>(&self, at: &BlockId<B::Block>, xts: T, force: bool)
+		-> Result<impl Future<Output=Vec<Result<ExHash<B>, B::Error>>>, B::Error>
+	where
+		T: IntoIterator<Item=ExtrinsicFor<B>>
+	{
+		let block_number = self.0.api.block_id_to_number(at)?
+			.ok_or_else(|| error::Error::InvalidBlockId(format!("{:?}", at)).into())?;
+
+		// for each xt, prepare a validation future
+		let validation_futures = xts.into_iter().map(move |xt| {
+			let (hash, bytes) = self.0.api.hash_and_length(&xt);
+			if !force && self.0.rotator.is_banned(&hash) {
+				return Either::Left(ready(Err(error::Error::TemporarilyBanned.into())))
+			}
+
+			let self_clone = self.clone();
+			Either::Right(self.0.api.validate_transaction(at, xt.clone())
+				.then(move |validation_result| ready(validation_result.and_then(|validity|
+					match validity {
+						Ok(validity) => if validity.provides.is_empty() {
+							Err(error::Error::NoTagsProvided.into())
+						} else {
+							Ok(base::Transaction {
+								data: xt,
+								bytes,
+								hash,
+								priority: validity.priority,
+								requires: validity.requires,
+								provides: validity.provides,
+								propagate: validity.propagate,
+								valid_till: block_number
+									.saturated_into::<u64>()
+									.saturating_add(validity.longevity),
+							})
+						},
+						Err(TransactionValidityError::Invalid(e)) => {
+							Err(error::Error::InvalidTransaction(e).into())
+						},
+						Err(TransactionValidityError::Unknown(e)) => {
+							self_clone.0.listener.write().invalid(&hash);
+							Err(error::Error::UnknownTransaction(e).into())
+						},
+					}
+				)))
+			)
+		});
+
+		// make single validation future that waits all until all extrinsics are validated
+		let validation_future = join_all(validation_futures);
+
+		// and then import all successfully validated extrinsics at once + enforce limits
+		let self_clone = self.clone();
+		let import_future = validation_future
+			.then(move |validation_results| ready({
+				let import_results = validation_results
+					.into_iter()
+					.map(|validation_result| {
+						let imported = self_clone.0.pool.write().import(validation_result?)?;
```

Comment:
This is locking inside the event loop :/ I think it might have very bad performance consequences.